### PR TITLE
[GPU] Fix PagedAttention intermediate buffer memory explosion

### PR DESCRIPTION
**Pull request by jade-cho**
### Description of the issue (symptom, root-cause, how it was resolved)
- **Symptom**: On ARL-HX (iGPU) Windows, PagedAttention models consume 3–8 GB extra device memory for intermediate buffers (`exp_sums`, `max_logits`, `tmp_output`), each ~90–235 MB × 36 layers × 3 buffers, all individually allocated without pool reuse.
- **Root cause**: The `paged_attention_node` constructor sets `can_share_internal_buffer(false)` as a blanket disable to prevent sharing of lockable buffers that use CPU `mem_lock<write>` in `prepare_internal_buffers()`. However, this also blocks pool sharing for GPU-only (non-lockable) intermediate buffers that are safe to share. (Introduced in PR #33204)
- **Resolution**: Removed the blanket `can_share_internal_buffer(false)` and instead made the pool-sharing decision per-buffer in `allocate_internal_buffer()`: lockable buffers (`m_lockable=true`) bypass the pool (preserving the original safety intent), while non-lockable buffers participate in `_non_padded_pool` sharing as designed.
- DG2 is unaffected because it uses `sdpa_micro_generate` kernel which does not allocate buf5/6/7.

#### The code and line that caused this issue (if it is not changed directly)
- `intel_gpu/src/graph/paged_attention.cpp` L16: `can_share_internal_buffer(false);`
- `intel_gpu/src/graph/impls/ocl_v2/sdpa/paged_attention_opt.cpp` `get_internal_buffer_descs()` — defines `m_lockable` per buffer descriptor

#### Problematic graph
- Any transformer model with 36+ PagedAttention layers where `pa_multi_tokens` kernel is selected (iGPU path, not `sdpa_micro_generate`)
- Each PA layer allocates ~10 internal buffers; buf5 (`exp_sums`), buf6 (`max_logits`), buf7 (`tmp_output`) are non-lockable GPU-only buffers that should be shared across layers

#### Checklist
- [x] Is it a proper fix? (not a workaround)
- [ ] Did you include test case for this fix, if necessary?
  - Unit test is not feasible: requires real GPU hardware, full PA topology compilation, and access to private `_intermediates_memory` / pool internals. No mock engine infrastructure exists. Verified correctness through static code analysis of the `_non_padded_pool` sharing path, `has_conflict()` / `basic_memory_dependencies` restriction sets, and `realloc_intermediates()` lifecycle.
- [x] Did you review existing test that can be extended to cover this scenario? Which test did you review?
  - Reviewed `memory_pool` tests in `memory_test.cpp` — tests output buffer pool sharing but not internal buffer lockable/non-lockable distinction.
  - Reviewed `paged_attention_gpu_test.cpp` (~5300 lines) — tests PA correctness/accuracy but not memory pool behavior. Existing PA tests serve as regression coverage for this change.

### Tickets:
 - *185192*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 1]
```diff
@@ -2492,7 +2492,7 @@ memory::ptr primitive_inst::allocate_internal_buffer(const layout& layout, size_
                              *_node,
                              layout,
                              alloc_type,
-                             can_share_internal_buffer(),
+                             can_share_internal_buffer() && !lockable,
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp]:
```suggestion
                             can_share_internal_buffer() || !lockable,
```
I think we need to disallow buffer sharing only when `can_share_internal_buffer` is false and `lockable` is true.

File: src/plugins/intel_gpu/src/graph/paged_attention.cpp [hunk 2]
```diff
@@ -13,7 +13,6 @@ GPU_DEFINE_PRIMITIVE_TYPE_ID(paged_attention)
 
 paged_attention_node::typed_program_node(const std::shared_ptr<paged_attention> prim, program& prog)
     : parent(prim, prog) {
-    can_share_internal_buffer(false);
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/paged_attention.cpp:16]: please keep this flag.

**Review by e-ddykim — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 1 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Changed

File: src/plugins/intel_gpu/src/graph/paged_attention.cpp (hunk 2 above)
Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/paged_attention.cpp:16]: Reverted

**Review by e-ddykim — APPROVED** [review]

**Comment by ceciliapeng2011** [comment]
The **internal_buffer** mechanism was originally designed for allocating temporary buffers. Extending it to support buffers with permanent residency may introduce additional complexity and potential issues. It would be helpful to revisit and clearly define the intended roles of the related flags, as well as how they are expected to interact with each other.


- The **lockable** flag appears to only describe the memory location of a buffer—whether it resides in usm_shared/usm_host or usm_device. From that perspective, the buffer should remain shareable regardless of the specific USM type.


- Using the **lockable** marker to bypass **can_share_internal_buffer** feels somewhat questionable and may not be fully justified. Since **share_internal_buffer** already exists, it might be cleaner and more consistent to rely on that flag instead.


- In addition, **share_internal_buffer** is currently applied at the node/primitive level, which seems overly coarse‑grained. A single primitive may contain multiple internal buffers, some of which are safe to share and others that are not. Treating all internal buffers uniformly under a single flag risks oversimplifying these cases and limiting flexibility.


- I also do not fully understand the original motivation for introducing **share_internal_buffer**, as referenced in “[GPU] adding sdpa_micro_gqa_single_token (#33204)”. Specifically, it is unclear why using **lockable** alone would result in a crash or hang in that context.


Given this, one possible approach could be:

1. If an internal buffer cannot be shared, allocate it from a dedicated memory pool like "const".
2. Otherwise, if allocation from a shared pool is required, keep **lockable** and **share_internal_buffer** semantics clearly separated. In particular, share_internal_buffer could be used solely to indicate whether a buffer is eligible for allocation from a shared pool.
3. Apply **share_internal_buffer** at the level of individual internal buffers rather than at the primitive level, allowing finer‑grained control and more accurate expression of sharing intent.

@riverlijunjie  @jade-cho  @e-ddykim

**Review by ceciliapeng2011 — APPROVED** [review]

**Review by riverlijunjie — APPROVED** [review]
